client/list.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2021 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+/*
+ * Module   : list.c
+ *
+ * Abstract :
+ *
+ *            tdnfclientlib
+ *
+ *            client library
+ *
+ * Authors  : Nitesh Kumar (kunitesh@vmware.com)
+ */
+
+#include "includes.h"
+
+TDNF_ML_LIST* 
+TDNFMergeList(
+    TDNF_ML_LIST* listA,
+    TDNF_ML_LIST* listB
+    )
+{
+    uint32_t dwError = 0;
+    TDNF_ML_LIST* mergedList = NULL;
+
+    // Base cases
+    if (listA == NULL) 
+        return (listB);
+    else if (listB == NULL)
+        return (listA);
+
+    //Compare the URL Preference to sort the URL List in descending order. 
+    TDNF_ML_URL_INFO *urlA = listA->data;
+    TDNF_ML_URL_INFO *urlB = listB->data;
+
+    if((urlA == NULL) || (urlB == NULL))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (urlA->preference >= urlB->preference) 
+    {
+        mergedList = listA;
+        mergedList->next = TDNFMergeList(listA->next, listB);
+    }
+    else 
+    {
+        mergedList = listB;
+        mergedList->next = TDNFMergeList(listA, listB->next);
+    }
+
+cleanup:
+    return (mergedList);
+
+error:
+    goto cleanup;
+}
+
+/* This function is used to split the nodes of the given
+ * list into two halves,and return the two lists using the
+ * reference. If the length is odd, the extra node should
+ * go in the front list.
+ */
+void 
+TDNFFrontBackSplit(
+    TDNF_ML_LIST* currHead,
+    TDNF_ML_LIST** frontRef,
+    TDNF_ML_LIST** backRef
+    )
+{
+    if((currHead == NULL) || (currHead->next == NULL))
+    {
+        return;
+    }
+
+    TDNF_ML_LIST* slowPtr = currHead; 
+    TDNF_ML_LIST* fastPtr = currHead->next;
+
+    while (fastPtr != NULL) 
+    {
+        fastPtr = fastPtr->next;
+        if (fastPtr != NULL) 
+        {
+            slowPtr = slowPtr->next;
+            fastPtr = fastPtr->next;
+        }
+    }
+
+    *frontRef = currHead;
+    *backRef = slowPtr->next;
+    slowPtr->next = NULL;
+}
+
+void 
+TDNFSortListOnPreference(
+    TDNF_ML_LIST** headUrl
+    )
+{
+    TDNF_ML_LIST* head = *headUrl;
```

Comment:
should check headUrl for NULL

Reply:
done

Reply:
Its getting checked through HEAD.

Reply:
if headUrl is NULL then dereferencing it would cause a crash. pls fix this.

Reply:
Done